### ISSUE TITLE
Schema comparator fails to detect addition of an optional field

### ISSUE THREAD

**Issue by seveneves** [softwaremill/sttp-apispec#170]
I would like to use `sttp.apispec.validation.SchemaComparator` to check if two versions of schema are compatible. But it fails to identify addition of a new optional field.

The following application highlights the problem

```scala
import sttp.apispec.validation.{SchemaComparator, SchemaCompatibilityIssue}
import sttp.tapir.Schema
import sttp.tapir.docs.apispec.schema.TapirSchemaToJsonSchema

object TestSchemas extends App {

  object v1 {

    case class TestSchema(oldField: String)
    object TestSchema {
      implicit val Schema: Schema[TestSchema] = sttp.tapir.Schema.derived[TestSchema]
    }
  }

  object v2 {

    case class TestSchema(oldField: String, newField: Option[String] = None)
    object TestSchema {
      implicit val Schema: Schema[TestSchema] = sttp.tapir.Schema.derived[TestSchema]
    }
  }

  private def apiSpecSchema[T: Schema] = TapirSchemaToJsonSchema(
    implicitly[Schema[T]],
    markOptionsAsNullable = true,
  )

  val v1Schema = apiSpecSchema[v1.TestSchema]
  val v2Schema = apiSpecSchema[v2.TestSchema]
  val comparator = new SchemaComparator(Map.empty, Map.empty)
  val issues: List[SchemaCompatibilityIssue] = comparator.compare(v2Schema, v1Schema)
  issues.foreach(println)
  assert(issues.isEmpty)
}
```

it produces 

>  schemas differ, but it was not possible to determine their compatibility

I read that `SchemaComparator` should support such cases 

>   In practice, the comparator is designed to detect typical changes that may appear in schemas during API evolution, e.g. adding new fields, changing types, etc.

But also it has the following 

> In more complex situations, the comparator simply falls back to comparing schemas by plain equality, or returns a GeneralSchemaMismatch

Which seems to be producing my output. 

So it is not clear if this is a complex case or it is a bug in `SchemaComparator`

**Comment by seveneves** [comment]
The problem seems to be caused by method `isProductSchema` which does the following check 

```scala
    s.`type`.getOrElse(Nil).filter(_ != SchemaType.Null) == List(SchemaType.Object) &&
      s.properties.nonEmpty &&
      s == Schema(
        `type` = s.`type`,
        properties = s.properties,
        required = s.required,
        dependentRequired = s.dependentRequired,
      )
```
And in my case last check is not fulfilled. It is caused by having  `$schema` set to `http://json-schema.org/draft-04/schema#`. So after removing it, the comparator works fine. 

I think `s == Schema` should be replaced with a better check that avoids problems like the one I just described.

**Comment by adamw** [comment]
Would a fix be to ignore the value of `$schema` on nested schemas? The [spec](https://json-schema.org/draft/2020-12/json-schema-core#section-8) says that embedded schemas MAY specify it, and I suppose schemas with and without `$schema` are equivalent, as this only influences potential validation of the schema itself, not of the schema's semantics?

Maybe @ghik will have something to say as well :)

**Comment by ghik** [comment]
It might be the case that stripping `$schema` should be added to [this method](https://github.com/softwaremill/sttp-apispec/blob/82c92be566ceb3b1f6642c43c383a4e55f825a26/apispec-model/src/main/scala/sttp/apispec/validation/SchemaComparator.scala#L102) - I don't recall exactly why it wasn't so far.

**Comment by seveneves** [comment]
This was fixed in #179
